### ISSUE TITLE
OpenID 2.0 implementation does not implement openid.identity=http://specs.openid.net/auth/2.0/identifier_select

### ISSUE THREAD

**Issue by francois-montmasson-efg** [Jasig/cas#930]
CAS 1.0 protocole is most similar to dumb OpenID authentication with openid.identity set with http://specs.openid.net/auth/2.0/identifier_select.

c.f. https://openid.net/specs/openid-authentication-2_0.html

Unfortunatly, this scenario is not supported by CAS 4.0 and CAS login dispay login page with identifier_select preselected username.

**Comment by leleuj** [comment]
Do you mean it was supported in CAS 3.5.x and doesn't work any more?

**Comment by francois-montmasson-efg** [comment]
Haven’t tested it in 3.5.x.

**Comment by leleuj** [comment]
Just for my understanding: if the `openid.identity` parameter is set to http://specs.openid.net/auth/2.0/identifier_select, the login field should not be predefined and could be entered by the user. Is this the only change in the OpenID mechanism?

**Comment by mmoayyed** [comment]
Found this might help:

openid.claimed_id
Value: (optional) The Claimed Identifier.

"openid.claimed_id" and "openid.identity" SHALL be either both present or both absent. If neither value is present, the assertion is not about an identifier, and will contain other information in its payload, using extensions.

openid.identity
Value: (optional) The OP-Local Identifier.

If a different OP-Local Identifier is not specified, the claimed identifier MUST be used as the value for openid.identity.

Note: If this is set to the special value "http://specs.openid.net/auth/2.0/identifier_select" then the OP SHOULD choose an Identifier that belongs to the end user. This parameter MAY be omitted if the request is not about an identifier (for instance if an extension is in use that makes the request meaningful without it; see openid.claimed_id above).

https://openid.net/specs/openid-authentication-2_0.html

**Comment by mmoayyed** [comment]
In my latest openid demo tryout, I did see the identifier was chosen by CAS automatically and all I had to do was type the password. I don't know what the value of the above parameter was though. @leleuj may know best.

**Comment by francois-montmasson-efg** [comment]
This parameter say we want the user to define his ID when he enters in CAS not before. This way user enter his login and password at the same time in cas and is not forced to enter it in client application before.


Regards,

Francois

**Comment by mmoayyed** [comment]
@leleuj Might this be something to include in 404?

**Comment by leleuj** [comment]
Yes, I don't think it's too much work.

**Comment by mmoayyed** [comment]
I do plan to get started with the release this Saturday. @leleuj do you still think that is enough time for you to close this issue, or shall we push back to 405?

**Comment by leleuj** [comment]
I'll see if I can do that tomorrow. If you don't see any PR for that this Saturday, we'll postpone for 4.0.5.

**Comment by mmoayyed** [comment]
Pushed to 4.0.5

**Comment by leleuj** [comment]
Things are more complicated than expected. I'm working on it.

**Comment by mmoayyed** [comment]
sure. thanks for the update

**Comment by leleuj** [comment]
@francois-montmasson-efg : Here are the current HTTP requests :

Going to the CAS server:

```
http://localhost:8080/cas/login?
openid.ns=http%3A%2F%2Fspecs.openid.net%2Fauth%2F2.0
&openid.claimed_id=http%3A%2F%2Flocalhost%3A8080%2Fcas%2Fopenid%2Fleleuj
&openid.identity=http%3A%2F%2Flocalhost%3A8080%2Fcas%2Fopenid%2Fleleuj
&openid.return_to=http%3A%2F%2Flocalhost%3A8080%2Fdemoo%2FopenId.html
&openid.realm=http%3A%2F%2Flocalhost%3A8080%2Fdemoo%2FopenId.html
&openid.assoc_handle=1439209853409-0
&openid.mode=checkid_setup
&openid.ns.sreg=http%3A%2F%2Fopenid.net%2Fsreg%2F1.0
&openid.sreg.optional=email%2Cfullname%2Cdob%2Cpostcode
```

Coming back to the application:

```
http://localhost:8080/demoo/openId.html?
openid.return_to=http%3A%2F%2Flocalhost%3A8080%2Fdemoo%2FopenId.html
&openid.identity=http%3A%2F%2Flocalhost%3A8080%2Fcas%2Fopenid%2Fleleuj
&openid.op_endpoint=http%3A%2F%2Flocalhost%3A8080%2Fcas%2Flogin
&openid.assoc_handle=1439298403594-0
&openid.mode=id_res
&openid.signed=op_endpoint%2Cclaimed_id%2Cidentity%2Creturn_to%2Cresponse_nonce%2Cassoc_handle
&openid.sig=%2BGHHr9KMC2mafkNhXfHqKhAut%2FmqRZ%2B8op9JXqSLF18%3D
&openid.claimed_id=http%3A%2F%2Flocalhost%3A8080%2Fcas%2Fopenid%2Fleleuj
&openid.response_nonce=2015-08-11T13%3A18%3A03Z0&openid.ns=http%3A%2F%2Fspecs.openid.net%2Fauth%2F2.0
```

I have questions with the new scenario:

- the `openid.identity` must be set to `http://specs.openid.net/auth/2.0/identifier_select` : what should be the value of the `openid.claimed_id`? Are other parameters changed?

- after a successfull authentication (say username: leleuj), what should be the values of the `openid.identity` and `openid.claimed_id`? Are other parameters changed?

**Comment by francois-montmasson-efg** [comment]
Hi,

Thanks for having a look in this issue. 

> the openid.identity must be set to http://specs.openid.net/auth/2.0/identifier_select : what should be the value of the openid.claimed_id?
> 


Setting both fields with this same value is an absolute requirement of the specification. c.f. 9.1

> Are other parameters changed?
> 
No

> after a successfull authentication (say username: leleuj), what should be the values of the openid.identity and openid.claimed_id? Are other parameters changed?
> 
There is no Claimed Identifier. c.f. 7.3.1
but 	OP-Local Identifier contains the authenticated user.

For more details, below are the extract from the Opend ID spec https://openid.net/specs/openid-authentication-2_0.html <https://openid.net/specs/openid-authentication-2_0.html> about http://specs.openid.net/auth/2.0/identifier_select <http://specs.openid.net/auth/2.0/identifier_select>:
 
7.3.1.  Discovered Information

Upon successful completion of discovery, the Relying Party will have one or more sets of the following information (see the Terminology section <https://openid.net/specs/openid-authentication-2_0.html#terminology> for definitions). If more than one set of the following information has been discovered, the precedence rules defined in [XRI_Resolution_2.0] <https://openid.net/specs/openid-authentication-2_0.html#XRI_Resolution_2.0> are to be applied.

OP Endpoint URL
Protocol Version
If the end user did not enter an OP Identifier, the following information will also be present:

Claimed Identifier
OP-Local Identifier
If the end user entered an OP Identifier, there is no Claimed Identifier. For the purposes of making OpenID Authentication requests, the value "http://specs.openid.net/auth/2.0/identifier_select" MUST be used as both the Claimed Identifier and the OP-Local Identifier when an OP Identifier is entered.

 <>
9.1.  Request Parameters

openid.claimed_id
Value: (optional) The Claimed Identifier.

"openid.claimed_id" and "openid.identity" SHALL be either both present or both absent. If neither value is present, the assertion is not about an identifier, and will contain other information in its payload, using extensions <https://openid.net/specs/openid-authentication-2_0.html#extensions>.

It is RECOMMENDED that OPs accept XRI identifiers with or without the "xri://" prefix, as specified in the Normalization <https://openid.net/specs/openid-authentication-2_0.html#normalization> section.

openid.identity
Value: (optional) The OP-Local Identifier.

If a different OP-Local Identifier is not specified, the claimed identifier MUST be used as the value for openid.identity. 

Note: If this is set to the special value "http://specs.openid.net/auth/2.0/identifier_select" then the OP SHOULD choose an Identifier that belongs to the end user. This parameter MAY be omitted if the request is not about an identifier (for instance if an extension is in use that makes the request meaningful without it; see openid.claimed_id above).


10.  Responding to Authentication Requests

If the relying party requested OP-driven identifier selection by setting "openid.identity" to "http://specs.openid.net/auth/2.0/identifier_select" and there are Identifiers for which the end user is authorized to issue authentication responses, the OP SHOULD allow the end user to choose which Identifier to use.


11.2.  Verifying Discovered Information

If the Claimed Identifier was not previously discovered by the Relying Party (the "openid.identity" in the request was "http://specs.openid.net/auth/2.0/identifier_select" or a different Identifier, or if the OP is sending an unsolicited positive assertion), the Relying Party MUST perform discovery on the Claimed Identifier in the response to make sure that the OP is authorized to make assertions about the Claimed Identifier.

14.2.1.  Relying Parties

In authentication requests <https://openid.net/specs/openid-authentication-2_0.html#requesting_authentication>, the "openid.identity" parameter SHOULD NOT be the special value "http://specs.openid.net/auth/2.0/identifier_select", because OpenID Authentication 1.1 does not support the use of OP Identifiers.

**Comment by leleuj** [comment]
Here is the PR: https://github.com/Jasig/cas/pull/1093

Feedbacks and tests would be appreciated. Before opening the same PR for 4.0.x

**Comment by leleuj** [comment]
Fixed in 4.1.
